### PR TITLE
Chrome: Enhance the publish button flow (and the saved state component)

### LINKED ISSUE #853: Sidebar: Link to latest revision
Design:

screen shot 2017-05-22 at 11 07 01

Clicking this link should simply take you to the latest revision of the post (example /wp-admin/revision.php?revision=23595). That page has buttons to navigate to previous revisions also. 

Context:

post settings

### PR DESCRIPTION
**Pull request by youknowriad**
closes #945  

Several things in this PR:

 - We need a way to match the server's timezone in the frontend in order to compare dates properly. I'm using moment-timezone for this and setting the server timezone globally in moment
 - When publishing a post (publish button), the status applied to the button 'publish' if no date or date in the past, 'future' if date in the future but this status can be overridden when the edits (pending, private, ...)
 - When clicking the "save" link, we always save a draft unless the post is already saved, we'll keep the same status in this case.

I'd appreciate help in testing this PR because there's so many use-cases. It's easy to miss one.

Auto-saving is not implemented yet.

### CODE REVIEW

**Comment by nylen** [comment]
@youknowriad: have you seen #920, which allows us to know whether the last save request was for a new draft (`Saved!`) or an existing post (`Updated!`)?  This is not quite working correctly in `master`.  I see that the related `requestIsNewPost` property was instead removed in this PR, and I am wondering how to reconcile these two PRs.

**Comment by youknowriad** [comment]
> I see that the related requestIsNewPost property was instead removed in this PR, and I am wondering 

I didn't remove it because it was broken but because it's not needed anymore. The button doesn't need to know if the last request was for an unsaved post or not.

**Comment by nylen** [comment]
If we don't need it, let's also remove that part of the state tree and the related action properties.  I wasn't very happy with how that looked anyway.

**Comment by jasmussen** [comment]
I totally love this. I feel like the experience completely vindicates the idea of separating the two. Thanks so much for working on this. 👍

**Comment by jasmussen** [comment]
Question: does auto-save ever kick in? If not, then how does that work, if at all? And if not, we should probably ticket this separately. 

In the context of this design, I imagine the behavior being like this:

1. Start new post. Left side area is empty
2. Type a little. Left side shows "Save".
3a. Wait a few seconds without typing, left side shows "Saving..." then "✔ Saved"
3b. Or, don't wait, click Save manually, and left side shows "Saving..." then "✔ Saved"

Autosave can totally be addressed separately. Additional future enhancements, could be an autosave error message when you're trying to save but the internet is down. This error message should show up on the left, like  "❗ Couldn't save [Try again]".

**Comment by youknowriad** [comment]
Worth noting there's a difficult to solve edge case in this PR: We can't have 

 - a "pending" and "draft" post at the same time because this information is stored in the same "status" field. So when hitting "Save" on a post where we activated the "pending" toggle. This will be reset since the post is saved as "draft".

@jasmussen This PR doesn't address auto-saving yet. I was planning to do this separately if the workflow here is validated (probably needs a separate issue). I'm not sure how to solve the issue above when auto-saving (Leaving the pending toggle activated and saving the post at the same time). The solution will be to save all the user changes except the "status" but this leaves the post as "dirty" which means the "Saved !" message will never appear. I'm thinking of showing this message for like 5 seconds after the save is successful (whether it's auto-save or explicit) and don't take into consideration the "dirty" property. Thoughts on this?

**Comment by jasmussen** [comment]
> @jasmussen This PR doesn't address auto-saving yet. I was planning to do this separately if the workflow here is validated (probably needs a separate issue). I'm not sure how to solve the issue above when auto-saving (Leaving the pending toggle activated and saving the post at the same time). The solution will be to save all the user changes except the "status" but this leaves the post as "dirty" which means the "Saved !" message will never appear. I'm thinking of showing this message for like 5 seconds after the save is successful (whether it's auto-save or explicit) and don't take into consideration the "dirty" property. Thoughts on this?

> Edit: We have the same issue with the "private" toggle.

@mtias @timmyc can you recall how we solved this in Calypso?

**Comment by youknowriad** [comment]
@jasmussen I looked at Calypso, and here's how it's solved:

- For private posts, toggling the "private" status publishes privately the post directly without another click on the "publish" button. This solves the issue, because the auto-save will just save the published post again without any status change.

- For pending posts, toggling the "pending" toggle don't save the post immediately but auto-saving and saving the post manually save the post as "pending" instead of "draft".

Does this sound ok for you?

**Comment by jasmussen** [comment]
That sounds pretty good to me, yes.

**Comment by youknowriad** [comment]
@jasmussen It should be ok now. I'm using `window.confirm` temporary instead of a Modal (we don't have that yet). I think we should ticket the Modal and the auto-saving separately. This PR is big enough, isn't it?

**Comment by jasmussen** [comment]
Nice, this is working and looking really well.

Visibility controls seem to have stopped working for me, though. Think it's this branch, or the webpack deduplication stuff?

**Comment by youknowriad** [comment]
> Visibility controls seem to have stopped working for me, though. Think it's this branch or the webpack deduplication stuff?

Good catch, it's fixed now by 1ab9d1e but I can't understand why the `onChange` handler is not enough. Really weird.

File: editor/assets/stylesheets/_animations.scss
Comment (aduth) [inline editor/assets/stylesheets/_animations.scss]: Tabbing inconsistency.

**Comment by aduth** [comment]
1. Navigate to Gutenberg
2. Click Save

Expected: "Publish" button labeled "Publish"
Actual: "Publish" button labeled "Schedule"

File: editor/header/saved-state/index.js
Comment (aduth) [inline editor/header/saved-state/index.js]: Should localize

File: editor/header/saved-state/index.js
Comment (aduth) [inline editor/header/saved-state/index.js]: Need for curly braces? `<Dashicon icon="saved" />`

File: editor/header/saved-state/index.js
Comment (aduth) [inline editor/header/saved-state/index.js]: Do we want this to be a block node, or is inline `span` better?

File: editor/header/saved-state/index.js
Comment (aduth) [inline editor/header/saved-state/index.js]: I wonder if block serialization should occur as part of a middleware before save proceeds into the network request handler, instead of the responsibility of the save button. What do you think?

File: editor/header/tools/publish-button.js
Comment (aduth) [inline editor/header/tools/publish-button.js]: There's no second argument accepted for `isEditedPostBeingScheduled`

File: editor/header/tools/style.scss
Comment (aduth) [inline editor/header/tools/style.scss]: Tabbing inconsistency. We need stylesheet linting 😄

File: editor/index.js
Comment (aduth) [inline editor/index.js]: Should this occur in `date/index.js` `setupLocale` ?

File: editor/state.js
Comment (aduth) [inline editor/state.js]: Do you think it be more semantic for edits to clear themselves by a standalone action `CLEAR_POST_EDITS`, which the effect handler(s) for `REQUEST_POST_UPDATE` would be responsible for returning/dispatching? Still unsure just how much we want to tie editor state to network activity, especially now that we have patterns for relatively simple effect chains.

File: editor/index.js
Comment (youknowriad) [inline editor/index.js]: The question is: if we do this in `data/index.js`, `wp-data` won't be library independent anymore. If you look at the code, you'll notice it's already done there, temporary but reverted to the original locale.

File: editor/header/saved-state/index.js
Comment (youknowriad) [inline editor/header/saved-state/index.js]: maybe. To be honest, sometimes I think we should avoid passing any edits, or content, or blocks here and use `getState` instead inside the effect handler. because it feels like I'm adding extra props to components when the component doesn't really need those props.

File: editor/state.js
Comment (youknowriad) [inline editor/state.js]: I understand this concern. And this is part of defining what is our Data Approach, how do we handle async actions etc... We should document this and be consistent.

File: editor/header/saved-state/index.js
Comment (aduth) [inline editor/header/saved-state/index.js]: >because it feels like I'm adding extra props to components when the component doesn't really need those props.

Yeah, that's more-or-less how I feel too. It doesn't seem to follow that the save button ought to be concerned with how to create the payload for the request. I think either an effect handler or a separate standalone middleware could serve in its place. I mentioned middleware only because I'm curious to see how far we can take effects to not rely on the store argument, per some of my own goals [outlined in the `refx` 2.0 changelog](https://github.com/aduth/refx/blob/master/CHANGELOG.md). I'd be fine with either approach, honestly, and the effect infrastructure being in place is a convenient lure.

File: editor/state.js
Comment (aduth) [inline editor/state.js]: I'll comment on #902

File: editor/header/tools/style.scss
Comment (aduth) [inline editor/header/tools/style.scss]: Still not quite tabbed enough 😅

File: editor/header/saved-state/index.js
Comment (aduth) [inline editor/header/saved-state/index.js]: Now that we're importing `__`, should we be consistent with how we're calling it throughout the file?

File: editor/header/tools/style.scss
Comment (aduth) [inline editor/header/tools/style.scss]: Can we override by exceeding specificity of those default styles, instead of through `!important` ? Why are we breaking consistency with default disabled button styling ? Should we be applying these overrides to `Button` more generally if they're warranted ?

File: editor/header/tools/style.scss
Comment (youknowriad) [inline editor/header/tools/style.scss]: 🙃

File: editor/header/saved-state/index.js
Comment (youknowriad) [inline editor/header/saved-state/index.js]: 🙃

File: editor/header/tools/style.scss
Comment (youknowriad) [inline editor/header/tools/style.scss]: Unfortunately we can't. These are the styles of WP and they are marked as `important` :(

File: editor/header/saved-state/index.js
Comment (youknowriad) [inline editor/header/saved-state/index.js]: I think it's probably better to leave this for a separate PR. The change doesn't concern this PR only.

**Comment by youknowriad** [comment]
@aduth I was able to fix the timezone issue, by creating a custom WP timezone for these custom timezones. 5f0f98

**Review by aduth — DISMISSED** [review]
Since we don't include `isNew` in `REQUEST_POST_UPDATE_SUCCESS`, the effect handler for navigating to apply `?post_id` is no longer triggered.

File: editor/index.js
Comment (aduth) [inline editor/index.js]: Is it independent this way though, if we're assigning defaults to the global instance?

File: editor/sidebar/post-status/index.js
Comment (aduth) [inline editor/sidebar/post-status/index.js]: 1. Navigate to Gutenberg demo page
2. Click "Pending review" twice (toggle on, then toggle off)
3. Click "Save"

Expected: A draft is saved
Actual: The post is published

Comment (aduth) [inline editor/sidebar/post-status/index.js]: Also, a few minor comments on this function since I found myself staring at it for a few minutes before figuring out what it's for:

- Naming of the function should probably reflect it's toggling related to pending status, e.g. `togglePendingStatus`
- Maybe some inline comments about the `if` behavior paths

Comment (aduth) [inline editor/sidebar/post-status/index.js]: Since this is a class, should we make this function an instance member?

File: editor/index.js
Comment (youknowriad) [inline editor/index.js]: It's independent in `wp-date` but an app like Gutenberg can make it global explicitly. That's my thinking. But I don't mind moving this to `wp-date` but we'd have to acknowledge that `wp-date` is moment dependent (externally).

**Review by aduth — DISMISSED** [review]
Minor unexpected flow (at least per existing editor behavior):

1. Navigate to Gutenberg
2. Publish post
3. Set date into future

Expected: Publish button reads "Schedule"
Actual: Publish button reads "Update"

File: editor/index.js
Comment (aduth) [inline editor/index.js]: I don't think we necessarily need to solve this here. Proposed for more discussion on the related Trac ticket: https://core.trac.wordpress.org/ticket/39222#comment:35

**Comment by jasmussen** [comment]
Excited about this branch! I'd like to do some responsive work. Should I do that in this branch, or wait until it's merged into master? I'm reluctant to do it in master currently, awaiting this one being merged in.

**Comment by jasmussen** [comment]
Actually, scratch that last comment. This branch already solves the responsive issues that are in master!

**Comment by youknowriad** [comment]
Yeah, let's try to merge this today. I think I've addressed all the concerns. I'll wait for another look

File: editor/effects.js
Comment (aduth) [inline editor/effects.js]: Question: Should we wait until after the save has completed to clear edits? What if the save fails?

File: editor/header/tools/publish-button.js
Comment (aduth) [inline editor/header/tools/publish-button.js]: Thinking some of this more complex logic warrants testing, or moving to selector e.g. `isPostPublishable`.

File: editor/selectors.js
Comment (aduth) [inline editor/selectors.js]: Question: Would this make more or less sense without "Already". I think it helps highlight that it tests whether the post was scheduled at a date before the current time, but maybe too much so to the extent that it might not be intuitive to use for a simple publish test.

File: editor/sidebar/post-status/index.js
Comment (aduth) [inline editor/sidebar/post-status/index.js]: What would it take to be able to show this for published posts as well? I'd suppose we need to revert back to the post's original status when toggling off, not a static value. Maybe an action that clears the current edited value for a property (we might want that for what we have anyways).

I'm not sure it makes much sense to revert from Published to Pending Review, but it is possible in the current editor.

Maybe if we at least have flow for reverting to draft it becomes a little more reasonable, since there's a workflow for changing from Published to Pending Review.

File: editor/effects.js
Comment (aduth) [inline editor/effects.js]: In fact, I notice something odd about this behavior. When changing visibility for a published post and Update-ing the post, the visibility reverts back to its original value briefly. I suspect it's because of this dispatch here.

**Review by aduth — COMMENTED** [review]
I left some comments but I think this is in a good state to merge if we want to iterate in subsequent pull requests.

File: editor/effects.js
Comment (youknowriad) [inline editor/effects.js]: Hard problem to solve, we should probably diff the edits before and after save to avoid loosing the unsaved edits (if we move this call in the onSuccess)

File: editor/sidebar/post-status/index.js
Comment (youknowriad) [inline editor/sidebar/post-status/index.js]: Good point, enabling it is not difficult to achieve, but yeah, it doesn't make much sense. I'm more in favour of having a "revert to draft" button here. cc @jasmussen

**Comment by youknowriad** [comment]
Thanks for the review @aduth I've addressed the selector notes. I'm going to merge this.
Here's a summary of the points to be addressed separately:

 - "Revert to draft" Button
 - Enhance the save effect to retrieve the edits and blocks from the store (as opposed to being passed as an argument)
 - Enhance the save effect to reset Edits on Save Success (and handle the edits made while saving)

Should we create separate issues for each of these points?

**Comment by jasmussen** [comment]
🎉

**Comment by aduth** [comment]
>"Revert to draft" Button

@jasmussen Do we have this in any of the other mockups?

@youknowriad Yeah, we should create separate issues. I'm hoping to take a look at a "save" middleware (re: second point) soon.

**Comment by jasmussen** [comment]
> Do we have this in any of the other mockups?

No specific mockups, no, but when we implemented this design we did for core, on WordPress.com first, we also designed that button there. Here's how it looks:

screen shot 2017-06-07 at 19 10 58

**Pull request merged**
